### PR TITLE
Chore: fix warning.

### PR DESCRIPTION
**Pull request by counter2015**
This pull request includes several changes to improve code quality, update dependencies, and enhance functionality. The most important changes include updating Scala versions and dependencies, adding new compiler options, and fixing various typos and formatting issues.

### Dependency Updates:
* Updated Scala versions to `3.3.3` and `2.12.19` in `build.mill`.
* Added `scala-collection-compat` dependency in `build.mill`.

### Code Enhancements:
* Added new compiler options (`-deprecation`, `-feature`, `-language:implicitConversions`, `-language:higherKinds`) in `build.mill`.
* Integrated `ScalafmtModule` into `OsModule` and added `collectionCompact` dependency in `build.mill`.

### Typo and Formatting Fixes:
* Corrected typos in documentation comments in `os/src/ListOps.scala` 
* Fixed various method signatures and added missing return types in `os/src/SubProcess.scala` 

### Migration to Modern Scala Collections:
* Replaced `collection.JavaConverters` with `scala.jdk.CollectionConverters` across multiple files (`os/src/Path.scala`, `os/src/ProcessOps.scala`, `os/src/ReadWriteOps.scala`, `os/src/ZipOps.scala`, `os/test/src-jvm/ZipOpJvmTests.scala`, `os/watch/src/WatchServiceWatcher.scala`)

### CODE REVIEW

File: build.mill [hunk 1]
```diff
@@ -26,6 +28,7 @@ object Deps {
   val sourcecode = ivy"com.lihaoyi::sourcecode::0.4.2"
   val utest = ivy"com.lihaoyi::utest::0.8.4"
   val expecty = ivy"com.eed3si9n.expecty::expecty::0.16.0"
+  val collectionCompact = ivy"org.scala-lang.modules::scala-collection-compat:2.9.0"
```
Comment (sake92) [inline build.mill:31]: Why is this needed?

Comment (counter2015) [reply build.mill:31]: Scala 2.12 can't import `scala.jdk.CollectionConverters` without this.

Comment (counter2015) [reply build.mill:31]: It seems have a higher version 2.12.0, let me update it.

**Comment by counter2015** [comment]
@sake92 is there anything else I can do to improve this PR ?

**Comment by lihaoyi** [comment]
I'm going to call this a wontfix. I don't find these improvements worth the churn they introduce to the codebase
